R/preprocessing.R

```diff
@@ -0,0 +1,152 @@
+##'   model that the data are to be used, e.g. [estimate_infections()] or
+##'   [estimate_secondary()]. See the documentation there for the expected
+##'   format.
+##' @param missing Character. Options are "missing" (the default), "accumulate"
+##'   and "zero". This determines how missing dates in the data are interpreted.
+##'   If set to "missing", any missing dates in the observation data will be
+##'   interpreted as missing and skipped in the likelihood. If set to
+##'   "accumulate", modelled observations will be accumulated and added to the
+##'   next non-missing data point. This can be used to model incidence data that
+##'   is reported at less than daily intervals. If set to "accumulate", the
+##'   first data point is not included in the likelihood (unless `initial` is
+##'   set to a non-zero value) but used only to reset modelled observations to
+##'   zero. If "zero" then all observations on missing dates will be assumed to
+##'   be of value 0.
+##' @param na Character. How to process dates that have NA values. The options
+##'   available are the same ones as for the `missing` argument.
+##' @param initial_accumulate Integer. The number of initial dates to accumulate
+##'   if `missing` or a column name is set to `"accumulate"`. This number of
+##'   dates is added to the beginning of the data set to be accumulated onto the
+##'   first data point. This is useful, for example, for modelling weekly
+##'   incidence data, in which case this should be set to 7. If accumulating and
+##'   the first data point is one to accumulate and this is not set, then that
+##'   data point will be removed with a warning.
+##' @param obs_column Character. If given, only the column specified here will
+##'   be used for checking missingness. This is useful if using a data set that
+##'   has NA values in multiple columns, but only one of them corresponds to
+##'   observations that are to be processed here. By default, all columns will
+##'   be processed and, e.g. `accumulate` will be set if any column is NA in any
+##'   given row.
+##' @return a data.table with an `accumulate` column that indicates whether
+##'   values are accumulated (see the documentation of the `data` argument in
+##'   [estimate_infections()])
+##' @importFrom rlang arg_match
+##' @importFrom data.table as.data.table data.table
+##' @export
+##' @examples
+##' cases <- data.table::copy(example_confirmed)
+##' ## calculate weekly sum
+##' cases[, confirm := data.table::frollsum(confirm, 7)]
+##' ## limit to dates once a week
+##' cases <- cases[seq(7, nrow(cases), 7)]
+##' ## set the second observation to missing
+##' cases[2, confirm := NA]
+##' ## fill missing data
+##' fill_missing(cases, missing = "accumulate", initial_accumulate = 7)
+fill_missing <- function(data,
+                         missing = c("missing", "accumulate", "zero"),
+                         na = c("missing", "accumulate", "zero"),
+                         initial_accumulate,
```

Comment:
I'm a little confused. I set `initial_accumulate = 1` and expected 2020-02-28  to be set to TRUE and an observation added at 2020-02-27. Is this expected?

```r
library(data.table)
cases <- data.table::copy(example_confirmed)
## calculate weekly sum
cases[, confirm := data.table::frollsum(confirm, 7)]
## limit to dates once a week
cases <- cases[seq(7, nrow(cases), 7)]
## set the second observation to missing
cases[2, confirm := NA]
## fill missing data
fill_missing(cases, missing = "accumulate", initial_accumulate = 1) |>
  head(14)

Key: <date>
          date confirm accumulate
        <Date>   <num>     <lgcl>
 1: 2020-02-28     647      FALSE
 2: 2020-02-29      NA       TRUE
 3: 2020-03-01      NA       TRUE
  
```

Reply:
Also, what should be the expected behavior if `intial_accumulate` is set to 0? Should it be the same as when it is not set?

Reply:
The idea is that the default is to accumulate one observation (`initial_accumulate = 1`). 

> 2020-02-28 to be set to TRUE and an observation added at 2020-02-27. 

This to me sounds like two observations are being accumulated (`initial_accumulate = 2`)

> Also, what should be the expected behavior if initial_accumulate is set to 0? Should it be the same as when it is not set?

We should probably make it have a minimum of 1.

Happy to hear of alternative/clearer suggestions for wording this.

Reply:
> We should probably make it have a minimum of 1.

Agreed. That can be added in the assertion on `initial_accumulate` I added below.

Reply:
I've updated the documentation in the attempt to make it clearer